server/src/test/java/org/opensearch/search/aggregations/metrics/CardinalityAggregatorTests.java

```diff
@@ -90,6 +98,56 @@ public void testRangeFieldValues() throws IOException {
         }, fieldType);
     }
 
+    public void testDynamicPruningOrdinalCollector() throws IOException {
+        final String fieldName = "testField";
+        final String filterFieldName = "filterField";
+
+        MappedFieldType fieldType = new KeywordFieldMapper.KeywordFieldType(fieldName);
+        final CardinalityAggregationBuilder aggregationBuilder = new CardinalityAggregationBuilder("_name").field(fieldName);
+        testAggregation(aggregationBuilder, new TermQuery(new Term(filterFieldName, "foo")), iw -> {
+            iw.addDocument(asList(
+                new KeywordField(fieldName, "1", Field.Store.NO),
+                new KeywordField(fieldName, "2", Field.Store.NO),
+                new KeywordField(filterFieldName, "foo", Field.Store.NO),
+                new SortedSetDocValuesField(fieldName, new BytesRef("1")),
+                new SortedSetDocValuesField(fieldName, new BytesRef("2"))
+            ));
+            iw.addDocument(asList(
+                new KeywordField(fieldName, "2", Field.Store.NO),
+                new KeywordField(filterFieldName, "foo", Field.Store.NO),
+                new SortedSetDocValuesField(fieldName, new BytesRef("2"))
+            ));
+            iw.addDocument(asList(
+                new KeywordField(fieldName, "1", Field.Store.NO),
+                new KeywordField(filterFieldName, "foo", Field.Store.NO),
+                new SortedSetDocValuesField(fieldName, new BytesRef("1"))
+            ));
+            iw.addDocument(asList(
+                new KeywordField(fieldName, "2", Field.Store.NO),
+                new KeywordField(filterFieldName, "foo", Field.Store.NO),
+                new SortedSetDocValuesField(fieldName, new BytesRef("2"))
+            ));
+            iw.addDocument(asList(
+                new KeywordField(fieldName, "3", Field.Store.NO),
+                new KeywordField(filterFieldName, "foo", Field.Store.NO),
+                new SortedSetDocValuesField(fieldName, new BytesRef("3"))
+            ));
+            iw.addDocument(asList(
+                new KeywordField(fieldName, "4", Field.Store.NO),
+                new KeywordField(filterFieldName, "bar", Field.Store.NO),
+                new SortedSetDocValuesField(fieldName, new BytesRef("4"))
+            ));
+            iw.addDocument(asList(
+                new KeywordField(fieldName, "5", Field.Store.NO),
+                new KeywordField(filterFieldName, "bar", Field.Store.NO),
+                new SortedSetDocValuesField(fieldName, new BytesRef("5"))
+            ));
+            }, card -> {
+            assertEquals(3.0, card.getValue(), 0);
```

Comment:
we should probably add an assertion on how many times collector.collect gets called, which should be 2 when dynamic pruning is applied vs 5 when its not applied?

Reply:
I have added test utilities (CountingAggregator) in https://github.com/opensearch-project/OpenSearch/pull/11643 with which we can assert the count of `collect()` calls.